packages/grid/_modules_/grid/hooks/features/columnReorder/useGridColumnReorder.tsx

```diff
@@ -103,27 +103,33 @@ export const useGridColumnReorder = (apiRef: GridApiRef): void => {
       const coordinates = { x: event.clientX, y: event.clientY };
 
       if (
-        params.field !== dragCol &&
+        params.field !== dragColField &&
         hasCursorPositionChanged(cursorPosition.current, coordinates)
       ) {
         const targetColIndex = apiRef.current.getColumnIndex(params.field, false);
-        const dragColIndex = apiRef.current.getColumnIndex(dragCol, false);
-
-        if (
-          (getCursorMoveDirectionX(cursorPosition.current, coordinates) ===
-            CURSOR_MOVE_DIRECTION_RIGHT &&
-            dragColIndex < targetColIndex) ||
-          (getCursorMoveDirectionX(cursorPosition.current, coordinates) ===
-            CURSOR_MOVE_DIRECTION_LEFT &&
-            targetColIndex < dragColIndex)
-        ) {
-          apiRef.current.setColumnIndex(dragCol, targetColIndex);
+        const targetColVisibleIndex = apiRef.current.getColumnIndex(params.field, true);
+        const targetCol = apiRef.current.getColumn(params.field);
+        const dragColIndex = apiRef.current.getColumnIndex(dragColField, false);
+        const visibleColumnAmount = apiRef.current.getVisibleColumns().length;
+
+        const canBeReordered =
+          !targetCol.disableReorder ||
+          (targetColVisibleIndex > 0 && targetColVisibleIndex < visibleColumnAmount - 1);
+
+        const cursorMoveDirectionX = getCursorMoveDirectionX(cursorPosition.current, coordinates);
+        const hasMovedLeft =
+          cursorMoveDirectionX === CURSOR_MOVE_DIRECTION_LEFT && targetColIndex < dragColIndex;
+        const hasMovedRight =
+          cursorMoveDirectionX === CURSOR_MOVE_DIRECTION_RIGHT && dragColIndex < targetColIndex;
+
+        if (canBeReordered && (hasMovedLeft || hasMovedRight)) {
+          apiRef.current.setColumnIndex(dragColField, targetColIndex);
         }
 
         cursorPosition.current = coordinates;
       }
     },
-    [apiRef, dragCol, logger],
+    [apiRef, dragColField, logger],
   );
 
   const handleDragEnd = React.useCallback(
```

Comment:
Small bug here. We need to check if the user started to drag the cell.

https://github.com/mui-org/material-ui-x/blob/cbd2179dc605d8f07bca15e60ef89ffe5fd600d4/packages/grid/_modules_/grid/hooks/features/columnReorder/useGridColumnReorder.tsx#L96-L98

![jLW0lPu6QF](https://user-images.githubusercontent.com/42154031/126001821-dd36a3e6-7fc0-4f9d-b7b8-c48323dfbb2a.gif)

Reply:
What is the issue on the video ?
I can check if `originColumnIndex.current` is not null.

Reply:
The age column is not reorderable but I select its text I can reorder. We just have to copy the same 3 lines from `handleDragOver`.